src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala

```diff
@@ -214,4 +228,31 @@ private[actions] abstract class CreateActionBase(dataManager: IndexDataManager)
     assert(partitionSchemas.length == 1)
     partitionSchemas.head.map(_.name)
   }
+
+  /**
+   * Generate a mapping for source data files by assigning a unique file id
+   * to each source data file. Once assigned, this file id does not change
+   * for a given file and is used to refer to that file.
+   *
+   * @param df DataFrame to index.
+   * @return mapping of source data file paths to their file ids, and the
+   *         last assigned file id.
+   */
+  protected def getFileIdsMap(df: DataFrame): (Map[String, Long], Long) = {
```

Comment:
Can we do this in sourceRelation (e.g. fromLeafFiles(allFiles, fileIdMap))?
I think it's better to assign `id` of FileInfo at construction - then we don't need `withFileIdsMap`

Reply:
For CreateIndex and RefreshIndex in the full mode, since we start id from 0, yes we can move the code there. However, I decided to keep this here to be consistent with IncrementalRefresh whose id assignment depends on the lastId from the log.
Moreover, to me, the fileIdMap population is an Action related task and I assumed it is better to keep it in the action code for clarity.  `fromLeafFiles` is a more general API and could be used outside of the Action context as well. I would prefer to keep this code here in the Action, but if you have strong preference to push it to `fromLeafFiles`, I will try to move it. Let me know @sezruby

Reply:
> For CreateIndex and RefreshIndex in the full mode, since we start id from 0, yes we can move the code there. However, I decided to keep this here to be consistent with IncrementalRefresh whose id assignment depends on the lastId from the log.

The last id can be collected from the previous log entry. 


> 
> Moreover, to me, the fileIdMap population is an Action related task and I assumed it is better to keep it in the action code for clarity.

Yes I also prefer the mutable fileIdMap in Action 


>  `fromLeafFiles` is a more general API and could be used outside of the Action context as well. 

File id is the part of FileInfo, so we can assign it when its construction. As 'fromLeafFiles' is an utility function, not the member function, I think we could define another utility function - fromLeafFilesWithFileId(files, map, last id)

> I would prefer to keep this code here in the Action, but if you have strong preference to push it to `fromLeafFiles`, I will try to move it. Let me know @sezruby 

I'd prefer to try this in this PR if possible.

Reply:
Can you plz confirm below is what you are asking for:
1. Remove `CreateActionBase.getFileNameToIdMap()`.
2. Make changes to `CreateActionBase.sourceRelations()` to assign ids to FileInfos. For this, add a new helper function to Content s.a `fromLeafFilesWithFileId(files, lastId): (Map[String, Long], Long)` which returns the Map of FileName -> id after creating FileInfos with Ids.

If above is true, then I think here is a bit of issue: This will work fine for Create and Refresh in the full mode, because they assign id to *all* source data files, from scratch. However, for RefreshIncremental, we should only assign id to *appended* files and preserve the id for any already existing file. By keeping "getFileNameToIdMap()" at the
action level , we can easily distinguish between these two sets of files. If we move this to `CreateActionBase.sourceRelations()` then I think we have to change its signature
to sth like `CreateActionBase.sourceRelations(df, map, lastId) where map is "previousIndexLogEntry.fileNameToIdMap" and lastId is `previousIndexLogEntry.lastFileId` so the existing files' ids get preserved. 

@sezruby  WDYT about above change? Do you still think it makes it better?

Reply:
It's not removing CreateActionBase.getFileNameToIdMap()
Keep it in CreateActionBase, but assign id when its creation.
```
// in CreateActionBase - you can keep them as a pair or separate variable. 
val fileIdMap = mutable.Map[String,Long] // this val can be overridden in RefreshAction with previousIndexLogEntry
var lastAssignedFileId = 0
// or keep the last id in fileIdMap 
def getNextId = { val nextId = fileIdMap.get(""), fileIdMap.put("", nextId + 1), nextId }

// in sourceRelation
...
val sourceDataProperties = Hdfs.Properties(Content.fromLeafFiles(files, fileidmap, lastid).get)
...
```

WDYT?

Reply:
Ok, I made changes to the code and now the id is assigned when the content is created from source.